### DISCUSSION TITLE
Quick Question about determinism in Godot Jolt

### DISCUSSION THREAD

**Discussion by Gamepro5** [godot-jolt/godot-jolt#548] [Questions] [answered]
In the README, you say:

> While Jolt itself offers deterministic simulations, Godot Jolt is not able to make such guarantees. Simulations in Godot Jolt may look deterministic, and may even happen to be deterministic, but this should not be relied upon if determinism is a hard requirement.

What exactly does determinism mean in this context? Google says it means that if ran on two separate machines, the outcome of the simulation will not be the same. How could this be? Does Godot Jolt have RNG in its algorithms? From all the MRPs I've produced, it seems that no matter what the bugs are able to be re-produced on different computers, so I would assume it is deterministic.

**Comment by mihe** [comment, marked as answer]
I'll preface by saying that determinism is not a topic I'm all that well versed in, which is part of why that disclaimer is there.

Deterministic simulations means that given the exact same input, meaning the order in which you step the simulation, the order in which you create bodies, the transforms that's set on a body, the force/torque that's applied to a body, etc. you will see the same simulation play out, down to the bit level.

There are multiple levels of determinism to consider. The first being that what I mentioned above plays out for the same compiled binary on the same platform. I would imagine this is probably the easier one to support from a physics engine standpoint. The only thing that can really trip you up here (the RNG you mentioned) would be if you introduce variable timesteps or threading, which is obviously not guaranteed to happen in the same order, so you need to make special considerations for that, which Jolt does internally.

Once you introduce different platforms, you typically also need to compile a different binary, maybe even with a different compiler. This opens you up to floating-point operations happening in different orders, depending on the compilation flags you use, or even optimizations that might shift some instructions around, creating ever so slightly different results that might look the same to the naked eye at first, but will eventually compound and break your determinism.

The `-ffast-math` stuff mentioned in #536 would be one example of such a compiler flag that is more or less guaranteed to break cross-platform determinism, since it deliberately trades in IEEE 754-compliance for performance, which is likely why that bug played out differently on Clang vs MSVC.

Anyway, there's more to this that I don't really feel qualified to talk about, but to sum up, deterministic simulations require very careful programming, and while it can certainly be done if you're integrating Jolt in your own game engine, where you have complete control of the inputs to the physics at all times, I just don't know how you would even begin making those kinds of sweeping guarantees with the amount of different features and settings that Godot offers.
